src/common_dims.cpp

```diff
@@ -0,0 +1,130 @@
+#include <migraphx/common_dims.hpp>
+#include <migraphx/ranges.hpp>
+#include <algorithm>
+#include <cassert>
+#include <numeric>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+template <class Iterator>
+static auto compute_end_dim(Iterator start, Iterator last, std::size_t dim)
+{
+    std::size_t x = 1;
+    auto it       = std::find_if(start, last, [&](auto i) {
+        x *= i;
+        return x > dim;
+    });
+    if(x < dim)
+        return start;
+    return it;
+}
+
+template <class Range>
+static auto elements(const Range& r)
+{
+    return std::accumulate(r.begin(), r.end(), std::size_t{1}, std::multiplies<>{});
```

Comment:
if begin == end then, it would still return 1 as the number of elements.

Reply:
> begin == end then 

`elements()` would never be called with begin==end. need to assert that.

Reply:
it is unlikely that this get called with `begin== end` but if you could add an assert for this, it would be good.